src/Microsoft.AspNet.Http/QueryCollection.cs

```diff
@@ -0,0 +1,412 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNet.Http.Internal
+{
+    /// <summary>
+    /// The HttpRequest query string collection
+    /// </summary>
+    public sealed class QueryCollection : IQueryCollection
```

Comment:
Manipulating the collection is problematic because it gets out of sync with request.QueryString. The recommended way to change something like Request.Query for downstream consumers is to update Request.QueryString, and then the QueryFeature will refresh. Similarly for cookies, you update the underlying headers. Path has no corresponding collection, just the raw PathString.

That said, I haven't seen any components need to re-write the query yet. Our auth middleware for example usually redirect if they've processed the request.